core/src/main/scala/org/http4s/HeaderV2.scala

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright 2013 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.header.v2
+
+import cats.{Id, Monoid, Order, Show}
+import cats.data.NonEmptyList
+import cats.syntax.all._
+import org.typelevel.ci.CIString
+import scala.collection.mutable.{ListBuffer, Set => MutSet}
+
+/** Typeclass representing an HTTP header, which all the http4s
+  * default headers satisfy.
+  * You can add custom headers by providing an implicit instance of
+  * `Header[YourCustomHeader]`
+  */
+trait Header[A, T <: Header.Type] {
+
+  /** Name of the header. Not case sensitive.
+    */
+  def name: CIString
+
+  /** Value of the header, which is represented as a String.
+    * Will be a comma separated String for headers with multiple values.
+    */
+  def value(a: A): String
+
+  /** Parses the header from its String representation.
+    * Could be a comma separated String in case of a Header with
+    * multiple values.
+    */
+  def parse(headerValue: String): Option[A]
```

Comment:
I'm wondering about `Option` vs. `Either`.  We've never told people why their headers are bad.

Reply:
Yeah that makes sense, although you then have to design an error type potentially

Reply:
We already have a `ParseFailure`, which is what all the header parsers return in a `ParseResult` alias.  I don't know that the latter alias carries its weight.  I was more pro-alias in those days.

Reply:
Are we definitely going for an `Either` then?

Reply:
I think it's a good idea to try.  If that makes the design irritating, copying back to `Option`, we can revert.

I think there's a rare opportunity for an `Ior` when combining headers when some of them are invalid, but maybe that's too cute, and I think we'd need a `Semigroup` for `ParseFailure`.

Reply:
:+1: for `Either`